### PR TITLE
ui: Switch bundling from create react app to webpack 5

### PR DESCRIPTION
**Pull request by JBWatenbergScality**
**Component**:

ui

**Context**: 

This allows us to have a greater control on our application bundling and enable access to advance webpack 5 configuration.
It also help us solving security vulnerabilities bringed by the old version of create react app we were constrained to use to hook into create react app buildchain.

**Acceptance criteria**: 

e2e tests

### CODE REVIEW

File: ui/README.md [hunk 1]
```diff
@@ -1,68 +1,50 @@
...
+```sh
+$ npm install
+```
 
-The page will reload if you make edits.<br>
-You will also see any lint errors in the console.
+## Run locally with a [metalk8s cluster already running](https://metal-k8s.readthedocs.io/en/latest/developer/running/cluster.html#)
 
-### `npm test`
+In `webpack.dev.js` edit the value of ´controlPlaneIP´ and provide your cluster bootstrap node ip. You can get it by running
```
Comment (gdemonet) [inline ui/README.md]:
```suggestion
In `webpack.dev.js` edit the value of `controlPlaneIP` and provide your cluster bootstrap node's control plane IP. You can get it by running:
```

File: ui/README.md [hunk 2]
```diff
@@ -1,68 +1,50 @@
...
 
-Builds the app for production to the `build` folder.<br>
-It correctly bundles React in production mode and optimizes the build for the best performance.
+```sh
+$ npm run start
+```
 
-The build is minified and the filenames include the hashes.<br>
-Your app is ready to be deployed!
+This will first starts `<root>/shell-ui` in docker container and expose it on the port 8084 and then will start the Metalk8s UI in dev mode.
```
Comment (gdemonet) [inline ui/README.md]:
```suggestion
This will first start `<root>/shell-ui` in a Docker container and expose it on port 8084 and then will start the Metalk8s UI in dev mode.
```

File: ui/README.md [hunk 3]
```diff
@@ -1,68 +1,50 @@
...
+If you don't have docker installed on your machine you can alternatively install `<root>/shell-ui` dependencies via running `npm install` in
```
Comment (gdemonet) [inline ui/README.md]:
```suggestion
If you don't have Docker installed on your machine you can alternatively install `<root>/shell-ui` dependencies via running `npm install` in 
```

File: ui/src/index.html [hunk 4]
```diff
@@ -11,16 +11,8 @@
...
-    <!--
-      Notice the use of %PUBLIC_URL% in the tags above.
-      It will be replaced with the URL of the `public` folder during the build.
-      Only files inside the `public` folder can be referenced from the HTML.
-
-      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
-      work correctly both with client-side routing and a non-root public URL.
-      Learn how to configure a non-root public URL by running `npm run build`.
-    -->
+    <link rel="manifest" href="manifest.json" />
```
Comment (gdemonet) [inline ui/src/index.html:14]: So we don't care about `%PUBLIC_URL%` anymore with Webpack 5? Or we just weren't using it anyway, and relying on Ingress to make it look like we're always served from a root URL?

File: ui/webpack.common.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      {
+        test: /\.html$/,
+        use: [
+          {
+            loader: 'html-loader',
+          },
+        ],
+      },
+      {
+        test: /\.jpe?g$|\.gif$|\.png$|\.ttf$|\.eot$|\.svg$/,
```
Comment (gdemonet) [inline ui/webpack.common.js]: Slightly more compact :wink:

```suggestion
        test: /\.(jpe?g|gif|png|ttf|eot|svg)$/,
```

File: ui/webpack.common.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            loader: 'file-loader',
+            options: {
+              name: '[name].[ext]?[hash]',
+              outputPath: 'static/media/',
+            },
+          },
+        ],
+      },
+      {
+        test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/,
```
Comment (gdemonet) [inline ui/webpack.common.js]: If this `?v=x.y.z` is supposed to be a version (SemVer?), I'm not sure this regex is flexible enough (what if `y` or `z` are greater than 9? what about pre-release identifiers?)

File: ui/README.md [hunk 7]
```diff
@@ -1,68 +1,50 @@
-This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
+# Metalk8s UI
 
-## Available Scripts
+## Prerequisite
```
Comment (gdemonet) [inline ui/README.md]:
```suggestion
## Prerequisites
```

File: ui/src/containers/CreateVolume.js [hunk 8]
```diff
@@ -376,8 +376,7 @@ const CreateVolume = (props) => {
       yup.object().shape({
         name: yup
           .string()
-          .matches(volumeNameRegex, intl.translate('name_error'))
-          .required(''),
+          .matches(volumeNameRegex, intl.translate('name_error')),
```
Comment (gdemonet) [inline ui/src/containers/CreateVolume.js:379]: Why is this required in your PR?

**Review by gdemonet — CHANGES_REQUESTED** [review]
Looks good to me at first sight, I'll give it a try :)
Some small suggestions for the README, although I think it would make more sense to document all of this under `docs/developer/running/ui.rst` (which is now outdated).

**Comment by gdemonet** [comment]
:disappointed: 

```sh
❯ npm install
npm ERR! code 1
npm ERR! git dep preparation failed
npm ERR! command /home/me/.nvm/versions/node/v14.16.1/bin/node /home/me/.nvm/versions/node/v14.16.1/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/home/me/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! npm ERR! code E401
npm ERR! npm ERR! Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
```

Any clue?

File: ui/src/index.html (hunk 4 above)
Comment (JBWatenbergScality) [reply ui/src/index.html:14]: We were relying on the ingress and never used the `%PUBLIC_URL%`. Additionally  `%PUBLIC_URL%` was injected at build time by create react app and requires an additional plugin not entirely ready for webpack 5. Also for the JS bundles the prefix is handled by the `output.publicPath` config entry

File: ui/src/containers/CreateVolume.js (hunk 8 above)
Comment (JBWatenbergScality) [reply ui/src/containers/CreateVolume.js:379]: The cypress tests were failing after moving to webpack 5. Likely a side effect of the dependencies bump I ran. The consequence was that the button remained disabled after having correctly filled the form because the volumes.name was not filled when creating a single volume.

**Comment by JBWatenbergScality** [comment]
> 😞
> 
> ```shell
> ❯ npm install
> npm ERR! code 1
> npm ERR! git dep preparation failed
> npm ERR! command /home/me/.nvm/versions/node/v14.16.1/bin/node /home/me/.nvm/versions/node/v14.16.1/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/home/me/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit
> npm ERR! npm WARN using --force Recommended protections disabled.
> npm ERR! npm ERR! code E401
> npm ERR! npm ERR! Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
> ```
> 
> Any clue?

I got this issue too with npm v7 few weeks ago (it is unrelated with webpack 5 changes). It seems one of our dependencies try to pull a dependency from a private npm repository hosted on AWS. For now I solved it by rollbacking npm to v6, but it is definitely something I'd like to solve.

File: ui/README.md [hunk 9]
```diff
@@ -1,68 +1,50 @@
-This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
+# Metalk8s UI
 
-## Available Scripts
+## Prerequisite 
 
-In the project directory, you can run:
+node v14+
+npm v6+
```
Comment (gdemonet) [inline ui/README.md]: So, given the issue with v7, maybe:
```suggestion
npm v6
```

File: ui/README.md (hunk 1 above)
Comment (JBWatenbergScality) [reply ui/README.md]: Done in 13f451b

File: ui/README.md (hunk 2 above)
Comment (JBWatenbergScality) [reply ui/README.md]: Done in 13f451b

File: ui/README.md (hunk 3 above)
Comment (JBWatenbergScality) [reply ui/README.md]: Done in 13f451b

File: ui/webpack.common.js (hunk 5 above)
Comment (JBWatenbergScality) [reply ui/webpack.common.js]: Done in 7279d56

File: ui/webpack.common.js (hunk 6 above)
Comment (JBWatenbergScality) [reply ui/webpack.common.js]: I removed the version as we are not using it. Done in 7279d56

File: ui/README.md (hunk 7 above)
Comment (JBWatenbergScality) [reply ui/README.md]: Done in 13f451b

File: ui/README.md (hunk 9 above)
Comment (JBWatenbergScality) [reply ui/README.md]: Done in 1b556a5

File: docs/developer/running/ui.rst [hunk 10]
```diff
@@ -20,33 +20,78 @@ Procedure
...
+    VERSION="your version (e.g. 2.9.1-dev)"
+    SALT_MASTER=$(kubectl \
+        --kubeconfig /etc/kubernetes/admin.conf get pods \
+        -n kube-system -l app=salt-master \
+        -o jsonpath='{.items[0].metadata.name}')
+    kubectl --kubeconfig /etc/kubernetes/admin.conf exec \
+        "$SALT_MASTER" -c salt-master -n kube-system -- \
+        salt-run state.sls metalk8s.addons.dex.deployed saltenv=metalk8s-$VERSION
+
+3. Enables CORS requests:
```
Comment (gdemonet) [inline docs/developer/running/ui.rst]: Small typo, and rewrite numbered lists using `#` instead of explicit numbers.
```suggestion
#. Enable CORS requests:
```

**Review by gdemonet — DISMISSED** [review]
Tested locally :ok_hand: (the bundle analyzer UI is cool :sunglasses:)

File: docs/developer/running/ui.rst (hunk 10 above)
Comment (JBWatenbergScality) [reply docs/developer/running/ui.rst]: Done in https://github.com/scality/metalk8s/pull/3390/commits/1f5fc92c1e23b6f4b5fad92bb2844d83bb819642 and rebased the branch

**Review by gdemonet — APPROVED** [review]
